sklearn/neighbors/nca.py

```diff
@@ -0,0 +1,526 @@
+
+    def _loss_grad_lbfgs(self, transformation, X, y, masks, sign=1.0):
+        """Compute the loss and the loss gradient w.r.t. ``transformation``.
+
+        Parameters
+        ----------
+        transformation : array, shape (n_features_out, n_features)
+            The linear transformation on which to compute loss and evaluate
+            gradient
+        X : array, shape (n_samples, n_features)
+            The training samples.
+
+        y : array, shape (n_samples,)
+            The corresponding training labels.
+
+        masks : array, shape (n_samples, n_classes)
+            One-hot encoding of y.
+
+        Returns
+        -------
+        loss : float
+            The loss computed for the given transformation.
+
+        gradient : array, shape (n_features_out * n_features,)
+            The new (flattened) gradient of the loss.
+        """
+
+        if self.n_iter_ == 0:
+            self.n_iter_ += 1
+            if self.verbose:
+                header_fields = ['Iteration', 'Objective Value', 'Time(s)']
+                header_fmt = '{:>10} {:>20} {:>10}'
+                header = header_fmt.format(*header_fields)
+                cls_name = self.__class__.__name__
+                print('[{}]'.format(cls_name))
+                print('[{}] {}\n[{}] {}'.format(cls_name, header,
+                                                cls_name, '-' * len(header)))
+
+        t_funcall = time.time()
+
+        transformation = transformation.reshape(-1, X.shape[1])
+        loss = 0
+        gradient = np.zeros(transformation.shape)
+        X_embedded = transformation.dot(X.T).T
+
+        # for every sample x_i, compute its contribution to loss and gradient
+        for i in range(X.shape[0]):
+            # compute squared distances to x_i in embedded space
+            diff_embedded = X_embedded[i] - X_embedded
+            dist_embedded = np.einsum('ij,ij->i', diff_embedded,
+                                      diff_embedded)
+            dist_embedded[i] = np.inf
+
+            # compute exponentiated distances (use the log-sum-exp trick to
+            # avoid numerical instabilities
+            exp_dist_embedded = np.exp(-dist_embedded -
+                                       logsumexp(-dist_embedded))
+            ci = masks[:, y[i]]  # samples that are in the same class as x_i
+            p_i_j = exp_dist_embedded[ci]
+            diffs = X[i, :] - X
+            diff_ci = diffs[ci, :]
+            diff_not_ci = diffs[~ci, :]
+            sum_ci = diff_ci.T.dot(
+                (p_i_j[:, np.newaxis] * diff_embedded[ci, :]))
+            sum_not_ci = diff_not_ci.T.dot((exp_dist_embedded[~ci][:,
+                                            np.newaxis] *
+                                            diff_embedded[~ci, :]))
+            p_i = np.sum(p_i_j)  # probability of x_i to be correctly
+            # classified
+            gradient += 2 * (p_i * (sum_ci.T + sum_not_ci.T) - sum_ci.T)
```

Comment:
You could just add the matrices without transposing, and transpose the gradient once before returning the unraveled gradient.

Reply:
I remembered I had seen a more efficient computation of the gradient (see last equation in slide 12 here: [http://bengio.abracadoudou.com/lce/slides/roweis.pdf](url) ). 
This would amount to:

`p_i_j = exp_dist_embedded[ci]`
`p_i_k = exp_dist_embedded[~ci]`
`diffs = X[i, :] - X`
`diff_ci = diffs[ci, :]`
`diff_not_ci = diffs[~ci, :]`
`sum_ci = diff_ci.T.dot(p_i_j[:, np.newaxis] * diff_ci)`
`sum_not_ci = diff_not_ci.T.dot(p_i_k[:, np.newaxis] * diff_not_ci)`
`p_i = np.sum(p_i_j)`
`gradient += p_i * sum_not_ci + (1 - p_i) * sum_ci`

And multiplying by the transformation after the for-loop:
`gradient = 2*np.dot(transformation, gradient)`

Reply:
> You could just add the matrices without transposing, and transpose the gradient once before returning the unraveled gradient.

True, I will change it.



> I remembered I had seen a more efficient computation of the gradient (see last equation in slide 12 here: http://bengio.abracadoudou.com/lce/slides/roweis.pdf ).
> This would amount to:
> [...]
> And multiplying by the transformation after the for-loop:
> gradient = 2*np.dot(transformation, gradient)

Since I already computed embedded differences for the softmax before wouldn't it be more efficient to reuse it ?
I agree that the expression with (1 - p_i) is clearer though.